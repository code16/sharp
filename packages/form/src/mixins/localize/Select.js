import localizeField from './field';

export default {
    mixins: [localizeField],

    methods: {
        localizeLabel(label) {
            return this.isLocalized ? label[this.locale] : label;
        },
        localizedOptionLabel(option) {
            return this.localizeLabel(option.label);
        }
    }
}