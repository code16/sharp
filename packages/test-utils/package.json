{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}