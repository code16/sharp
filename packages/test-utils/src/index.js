export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}