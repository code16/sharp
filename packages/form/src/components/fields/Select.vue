<template>
    <div class="SharpSelect" :class="[{'SharpSelect--multiple':multiple}, `SharpSelect--${display}`]">
        <Multiselect
            v-if="display==='dropdown'"
            :value="value"
            :searchable="false"
            :options="multiselectOptions"
            :multiple="multiple"
            :hide-selected="multiple"
            :close-on-select="!multiple"
            :custom-label="multiselectLabel"
            :placeholder="placeholder"
            :disabled="readOnly"
            :max="maxSelected"
            :allow-empty="clearable"
            @input="handleInput"
            @open="$emit('open')"
            @close="$emit('close')"
            ref="multiselect"
        >
            <template v-if="hasClearButton" v-slot:caret>
                <button class="SharpSelect__clear-button" type="button" @mousedown.stop.prevent="remove()">
                    <svg class="SharpSelect__clear-button-icon"
                         aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd">
                        <path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
                    </svg>
                </button>
            </template>
            <template v-slot:tag="{ option, remove }">
                <span class="multiselect__tag" :key="option">
                    <span>{{ multiselectLabel(option) }}</span>
                    <i aria-hidden="true" tabindex="1" @keypress.enter.prevent="remove(option)" @mousedown.prevent.stop="remove(option)" class="multiselect__tag-icon"></i>
                </span>
            </template>
            <template v-slot:option>
                <slot name="option" />
            </template>
        </Multiselect>
        <template v-else>
            <div class="SharpSelect__group" :class="{'SharpSelect__group--block':!inline}">
                <template v-if="multiple">
                    <template v-for="option in options">
                        <div class="SharpSelect__item" :class="itemClasses" :key="option.id">
                            <Check
                                :value="isChecked(option)"
                                :text="optionsLabel[option.id]"
                                :read-only="readOnly"
                                @input="handleCheckboxChanged($event, option)"
                            />
                        </div>
                    </template>
                </template>
                <template v-else>
                    <template v-for="(option, index) in options">
                        <div class="SharpSelect__item" :class="itemClasses" :key="option.id">
                            <input type="radio"
                                class="SharpRadio"
                                tabindex="0"
                                :id="`${uniqueIdentifier}${index}`"
                                :checked="isSelected(option)"
                                :value="option.id"
                                :disabled="readOnly"
                                :name="uniqueIdentifier"
                                @change="handleRadioChanged(option)"
                            >
                            <label class="SharpRadio__label" :for="`${uniqueIdentifier}${index}`">
                                <span class="SharpRadio__appearance"></span>
                                {{ optionsLabel[option.id] }}
                            </label>
                        </div>
                    </template>
                </template>
            </div>
        </template>
    </div>
</template>

<script>
    import { Multiselect } from 'sharp-ui';
    import Check from './Check.vue';
    import localize from '../../mixins/localize/Select';
    import { setDefaultValue } from "../../util";

    export default {
        name: 'SharpSelect',

        mixins: [localize],

        components: {
            Multiselect,
            Check
        },

        props: {
            value: [Array, String, Number],
            uniqueIdentifier: String,
            options: {
                type: Array,
                required: true,
                default: ()=>[],
            },
            multiple: {
                type: Boolean,
                default: false
            },
            display: {
                type: String,
                default: 'dropdown'
            },
            clearable: {
                type: Boolean,
                default: false
            },
            placeholder: {
                type: String,
                default: '-'
            },
            maxSelected: Number,
            readOnly: Boolean,

            inline: {
                type: Boolean,
                default: true
            },
        },

        data() {
            return {
                checkboxes: this.value
            }
        },
        watch: {
            options() {
                this.init();
            }
        },
        computed: {
            multiselectOptions() {
                return this.options.map(o => o.id);
            },
            optionsLabel() {
                // if (this.display !== 'dropdown')
                //     return;

                return this.options.reduce((map, opt) => {
                    map[opt.id] = this.localizedOptionLabel(opt);
                    return map;
                }, {});
            },
            itemClasses() {
                return {
                    'SharpSelect__item--inline': this.inline,
                }
            },
            hasClearButton() {
                return this.clearable && !this.multiple && this.value != null;
            },
        },
        methods: {
            isSelected(option, value = this.value) {
                if(option.id == null || value == null) {
                    return false;
                }
                return `${option.id}` === `${value}`;
            },
            isChecked(option) {
                return this.value?.some(value => this.isSelected(option, value));
            },
            remove() {
                this.$emit('input', null);
            },
            multiselectLabel(id) {
                return this.optionsLabel[id];
            },
            handleInput(val) {
                this.$emit('input', val);
            },
            handleCheckboxChanged(checked, option) {
                if (checked) {
                    this.$emit('input', [...(this.value ?? []), option.id])
                }
                else {
                    this.$emit('input', (this.value ?? []).filter(val => !this.isSelected(option, val)));
                }
            },
            handleRadioChanged(option) {
                this.$emit('input', option.id);
            },
            setDefault() {
                if(!this.clearable && this.value == null && this.options.length > 0) {
                    this.$emit('input', this.options[0].id, { force:true });
                }
            },
            init() {
                setDefaultValue(this, this.setDefault, {
                    dependantAttributes: ['options'],
                });
            }
        },
        created() {
            this.init();
        }
    }
</script>