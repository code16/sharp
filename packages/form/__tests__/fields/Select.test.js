import Vue from 'vue';
import Select from '../../src/components/fields/Select.vue';

import { MockI18n } from '@sharp/test-utils';



describe('select-field',()=>{
    Vue.use(MockI18n);

    beforeEach(()=>{
        document.body.innerHTML = `
            <div id="app">
                <sharp-select :value="value" 
                              :multiple="multiple" 
                              :display="display" 
                              :options="[{id:3,label:'AAA'}, {id:4, label:'BBB'}]" 
                              :read-only="readOnly"
                              :max-selected="multiple ? 3 : undefined"
                              placeholder="placeholder"
                              unique-identifier="select"
                              clearable
                              @input="inputEmitted($event)">
                </sharp-select>
            </div>
        `;
        MockI18n.mockLangFunction();
    });

    test('can mount Select field', async () => {
        await createVm({
            data:() => ({ value: null }) // possible String/Number value (no reactive)
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount Select field with clear button', async () => {
        await createVm({
            propsData: {
                multiple: false
            },
            data:() => ({ value: 3 }) // possible String/Number value (no reactive)
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount Select field as checkboxes list', async () => {
        await createVm({
            propsData: {
                multiple: true,
                display: 'list'
            },
            data: () => ({ value: [] })
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "read only" Select field as checkboxes list', async () => {
        await createVm({
            propsData: {
                multiple: true,
                display: 'list',
                readOnly: true,
            },
            data:() => ({ value: [] })
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount Select field as radios list', async () => {
        await createVm({
            propsData: {
                multiple: false,
                display: 'list'
            },
            data: () => ({ value: null })
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('can mount "read only" Select field as radios list', async () => {
        await createVm({
            propsData: {
                multiple: false,
                display: 'list',
                readOnly: true
            },
            data: () => ({ value: null })
        });
        expect(document.body.innerHTML).toMatchSnapshot();
    });

    test('expose appropriate props to multiselect component', async () => {

        let $tags = await createVm({
            propsData: {
                multiple: true, readOnly: true
            },
            data: () => ({ value: [3] })
        });

        let { multiselect } = $tags.$refs;

        expect(multiselect.$props).toMatchObject({
            value: [3],
            placeholder: 'placeholder',
            max: 3,
            disabled: true,
            searchable: false,
            // multiple dependant props
            closeOnSelect: false,
            multiple: true,
            hideSelected: true,
        });
    });

    test('expose appropriate props to multiselect component when multiple is false', async () => {

        let $tags = await createVm({
            propsData: {
                multiple: false
            },
            data: () => ({ value: null })
        });

        let { multiselect } = $tags.$refs;

        expect(multiselect.$props).toMatchObject({
            // multiple dependant props
            closeOnSelect: true,
            multiple: false,
            hideSelected: false,
        });
    });

    test('clear on cross button clicked', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            propsData: {
                multiple: false
            },
            methods: {
                inputEmitted
            },
            data: () => ({ value: 3 })
        });

        let clearBtn = document.querySelector('.SharpSelect__clear-button');
        clearBtn.dispatchEvent(new MouseEvent('mousedown'));

        expect(inputEmitted.mock.calls[0][0]).toBe(null);
    });

    test('check correct checkboxes depending on value', async () => {
        let value = [];
        await createVm({
            propsData: {
                multiple: true,
                display: 'list'
            },
            data:() => ({ value })
        });

        let checkboxes = document.querySelectorAll('input');

        expect(document.querySelectorAll('input:checked').length).toBe(0);

        value.push(3);
        await Vue.nextTick();

        expect(checkboxes[0].checked).toBe(true);
        expect(checkboxes[1].checked).toBe(false);
    });

    test('emit input on checkbox changed and correct value', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            propsData: {
                multiple: true,
                display: 'list'
            },
            methods: {
                inputEmitted
            },
            data: () => ({ value: [] })
        });

        let checkbox = document.querySelector('input');

        checkbox.click();

        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith([3]);
    });

    test('check correct radio depending on value', async () => {

        let { $root:vm } = await createVm({
            propsData: {
                multiple: false,
                display: 'list'
            },
            data: () => ({ value: null }),
        });

        let radios = document.querySelectorAll('input');

        expect(document.querySelectorAll('input:checked').length).toBe(0);

        vm.value = 3;
        await Vue.nextTick();

        expect(radios[0].checked).toBe(true);
        expect(radios[1].checked).toBe(false);
    });

    test('emit input on radio clicked and correct value', async () => {
        let inputEmitted = jest.fn();

        await createVm({
            propsData: {
                multiple: false,
                display: 'list'
            },
            methods: {
                inputEmitted
            },
            data: () => ({ value: null })
        });

        let radio = document.querySelector('input');

        radio.dispatchEvent(new Event('change', { bubbles: true }));

        expect(inputEmitted).toHaveBeenCalledTimes(1);
        expect(inputEmitted).toHaveBeenCalledWith(3);
    });

    test('correct options labels', async () => {
        let $select = await createVm({
            data: () => ({ value: null })
        });

        expect($select.multiselectLabel(3)).toBe('AAA');
        expect($select.multiselectLabel(4)).toBe('BBB');
    });

    test('corresponding multiselect options ids', async () => {
        let $select = await createVm({
            data: () => ({ value: null })
        });

        expect($select.multiselectOptions).toEqual([3,4]);
    });

    test('corresponding input id and label for', async () => {
        await createVm({
            propsData: {
                multiple:false,
                display:'list'
            },
            data: () => ({ value: null })
        });

        let radio = document.querySelector('input'),
            label = document.querySelector('label');

        expect(radio.id).toBe('select0');
        expect(label.htmlFor).toBe('select0');
    });
});

async function createVm(customOptions={}) {

    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],

        components: {
            'sharp-select': Select
        },

        props:['multiple', 'display', 'readOnly'],
    });

    await Vue.nextTick();

    return vm.$children[0];
}