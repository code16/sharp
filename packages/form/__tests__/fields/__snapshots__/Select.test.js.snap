// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select-field can mount "read only" Select field as checkboxes list 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--multiple SharpSelect--list">
		<div class="SharpSelect__group">
			<div class="SharpSelect__item SharpSelect__item--inline">
				<div class="SharpCheck">
					<label class="SharpCheck__label">
						<input type="checkbox" disabled="disabled" class="SharpCheck__input">
						<span class="SharpCheck__appearance">
							<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
								<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
							</svg>
						</span>
						<span class="SharpCheck__label-text">AAA</span>
					</label>
				</div>
			</div>
			<div class="SharpSelect__item SharpSelect__item--inline">
				<div class="SharpCheck">
					<label class="SharpCheck__label">
						<input type="checkbox" disabled="disabled" class="SharpCheck__input">
						<span class="SharpCheck__appearance">
							<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
								<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
							</svg>
						</span>
						<span class="SharpCheck__label-text">BBB</span>
					</label>
				</div>
			</div>
		</div>
	</div>
</div>
`;

exports[`select-field can mount "read only" Select field as radios list 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--list">
		<div class="SharpSelect__group">
			<div class="SharpSelect__item SharpSelect__item--inline">
				<input type="radio" tabindex="0" id="select0" disabled="disabled" name="select" class="SharpRadio" value="3">
				<label for="select0" class="SharpRadio__label">
					<span class="SharpRadio__appearance"></span>
					AAA
				</label>
			</div>
			<div class="SharpSelect__item SharpSelect__item--inline">
				<input type="radio" tabindex="0" id="select1" disabled="disabled" name="select" class="SharpRadio" value="4">
				<label for="select1" class="SharpRadio__label">
					<span class="SharpRadio__appearance"></span>
					BBB
				</label>
			</div>
		</div>
	</div>
</div>
`;

exports[`select-field can mount Select field 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--dropdown">
		<div tabindex="0" class="multiselect">
			<svg width="10" height="5" viewBox="0 0 10 5" fill-rule="evenodd" class="multiselect__select">
				<path d="M10 0L5 5 0 0z"></path>
			</svg>
			<div class="multiselect__tags">
				<div class="multiselect__tags-wrap" style="display: none;"></div>
				<!---->
				<div class="multiselect__spinner" style="display: none;"></div>
				<!---->
				<!---->
				<span class="multiselect__placeholder">
					placeholder
				</span>
			</div>
			<div tabindex="-1" class="multiselect__content-wrapper" style="max-height: 300px; display: none;">
				<ul class="multiselect__content" style="display: inline-block;">
					<!---->
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option">
							<span>AAA</span>
						</span>
						<!---->
					</li>
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option">
							<span>BBB</span>
						</span>
						<!---->
					</li>
					<li style="display: none;">
						<span class="multiselect__option">No elements found. Consider changing the search query.</span>
					</li>
					<li style="display: none;">
						<span class="multiselect__option">List is empty.</span>
					</li>
				</ul>
			</div>
		</div>
	</div>
</div>
`;

exports[`select-field can mount Select field as checkboxes list 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--multiple SharpSelect--list">
		<div class="SharpSelect__group">
			<div class="SharpSelect__item SharpSelect__item--inline">
				<div class="SharpCheck">
					<label class="SharpCheck__label">
						<input type="checkbox" class="SharpCheck__input">
						<span class="SharpCheck__appearance">
							<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
								<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
							</svg>
						</span>
						<span class="SharpCheck__label-text">AAA</span>
					</label>
				</div>
			</div>
			<div class="SharpSelect__item SharpSelect__item--inline">
				<div class="SharpCheck">
					<label class="SharpCheck__label">
						<input type="checkbox" class="SharpCheck__input">
						<span class="SharpCheck__appearance">
							<svg width="12" height="9" viewBox="0 0 12 9" fill-rule="evenodd" class="SharpCheck__checkmark">
								<path d="M4.1 6.1L1.4 3.4 0 4.9 4.1 9l7.6-7.6L10.3 0z"></path>
							</svg>
						</span>
						<span class="SharpCheck__label-text">BBB</span>
					</label>
				</div>
			</div>
		</div>
	</div>
</div>
`;

exports[`select-field can mount Select field as radios list 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--list">
		<div class="SharpSelect__group">
			<div class="SharpSelect__item SharpSelect__item--inline">
				<input type="radio" tabindex="0" id="select0" name="select" class="SharpRadio" value="3">
				<label for="select0" class="SharpRadio__label">
					<span class="SharpRadio__appearance"></span>
					AAA
				</label>
			</div>
			<div class="SharpSelect__item SharpSelect__item--inline">
				<input type="radio" tabindex="0" id="select1" name="select" class="SharpRadio" value="4">
				<label for="select1" class="SharpRadio__label">
					<span class="SharpRadio__appearance"></span>
					BBB
				</label>
			</div>
		</div>
	</div>
</div>
`;

exports[`select-field can mount Select field with clear button 1`] = `
<div id="app">
	<div class="SharpSelect SharpSelect--dropdown">
		<div tabindex="0" class="multiselect">
			<button type="button" class="SharpSelect__clear-button">
				<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpSelect__clear-button-icon">
					<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
				</svg>
			</button>
			<div class="multiselect__tags">
				<div class="multiselect__tags-wrap" style="display: none;"></div>
				<!---->
				<div class="multiselect__spinner" style="display: none;"></div>
				<!---->
				<span class="multiselect__single">AAA</span>
				<!---->
			</div>
			<div tabindex="-1" class="multiselect__content-wrapper" style="max-height: 300px; display: none;">
				<ul class="multiselect__content" style="display: inline-block;">
					<!---->
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option multiselect__option--selected">
							<span>AAA</span>
						</span>
						<!---->
					</li>
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option">
							<span>BBB</span>
						</span>
						<!---->
					</li>
					<li style="display: none;">
						<span class="multiselect__option">No elements found. Consider changing the search query.</span>
					</li>
					<li style="display: none;">
						<span class="multiselect__option">List is empty.</span>
					</li>
				</ul>
			</div>
		</div>
	</div>
</div>
`;
